frameworks/Rust/actix/src/db_pg_direct.rs

```diff
@@ -0,0 +1,154 @@
+            .into_future()
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e.0))
+            .and_then(|(row, _)| {
+                let row = row.unwrap();
+                Ok(World {
+                    id: row.get(0),
+                    randomnumber: row.get(1),
+                })
+            })
+    }
+
+    pub fn get_worlds(&self, num: usize) -> impl Future<Item=Vec<World>, Error=io::Error> {
+        let mut worlds = Vec::with_capacity(num);
+        for _ in 0..num {
+            let w_id: i32 = thread_rng().gen_range(1, 10_000);
+            worlds.push(
+                self.cl
+                    .query(&self.world, &[&w_id])
+                    .into_future()
+                    .map_err(|e| io::Error::new(io::ErrorKind::Other, e.0))
+                    .and_then(|(row, _)| {
+                        let row = row.unwrap();
+                        Ok(World {
+                            id: row.get(0),
+                            randomnumber: row.get(1),
+                        })
+                    }),
+            );
+        }
+
+        stream::futures_unordered(worlds).collect()
+    }
+
+    pub fn update(&self, num: usize) -> impl Future<Item=Vec<World>, Error=io::Error> {
+        let mut worlds = Vec::with_capacity(num);
+        for _ in 0..num {
+            let id: i32 = thread_rng().gen_range(1, 10_000);
+            let w_id: i32 =thread_rng().gen_range(1, 10_000);
+            worlds.push(
+                self.cl
+                    .query(&self.update, &[&w_id])
+                    .into_future()
+                    .map_err(|e| io::Error::new(io::ErrorKind::Other, e.0))
+                    .and_then(move |(row, _)| {
+                        let row = row.unwrap();
+                        Ok(World {
+                            id: row.get(0),
+                            randomnumber: id,
+                        })
+                    }),
+            );
+        }
+
+        let cl = self.cl.clone();
+        stream::futures_unordered(worlds)
+            .collect()
+            .and_then(move |mut worlds| {
+                let mut update = String::with_capacity(120 + 6 * num as usize);
+                update
+                    .push_str("UPDATE world SET randomnumber = temp.randomnumber FROM (VALUES ");
+
+                for w in &worlds {
+                    update.push_str(&format!("({}, {}),", w.id, w.randomnumber));
+                }
+                worlds.sort_by_key(|w| w.id);
+
+                update.pop();
+                update
+                    .push_str(" ORDER BY 1) AS temp(id, randomnumber) WHERE temp.id = world.id");
+
+                cl.batch_execute(&update)
+                  .map_err(|e| io::Error::new(io::ErrorKind::Other, e))
+                  .and_then(|_| Ok(worlds))
+            })
+    }
+
+    pub fn tell_fortune(&self) -> impl Future<Item=Vec<Fortune>, Error=io::Error> {
+        let mut items = Vec::with_capacity(16);
```

Comment:
The `16` here looks suspicious.  As a reminder, this is one of our fortunes test requirements:

> The list data structure must be a dynamic-size or equivalent and should not be dimensioned using foreknowledge of the row-count of the database table.

Could you change this so that you do not specify an initial capacity?  Would `Vec::new()` work?

Reply:
done. Vec::with_capacity() is just pre-allocation in heap.